### PR TITLE
CRAB-34138: Create Excel Spreadsheet with Resources and Upload to S3

### PR DESCRIPTION
**Pull request by alizbay94**
This PR allows NagBot to create a formatted excel spreadsheet containing the data pertaining to each of the Resources it processes, and the spreadsheet includes a worksheet for each resource. The spreadsheet is generated during the `notify` command, and it is created using the `xlsxwriter` library. 

The spreadsheet is uploaded to the [nagbot-spreadsheets](https://s3.console.aws.amazon.com/s3/buckets/nagbot-spreadhseets?region=us-east-1&tab=objects) S3 bucket in the main Seeq AWS account, and the format for naming the spreadsheets is `YYYY-MM-DD-NagBot-Report.xlsx`.

Here is an example spreadsheet that was produced: 
[2022-11-08-NagBot-Report.xlsx](https://github.com/seeq12/nagbot/files/9963705/2022-11-08-NagBot-Report.xlsx)

I've verified the changes work when running in a Docker container, and on my machine.

One thing I was unsure of:
Do we need to cleanup the excel file after uploading? I think the file should be deleted at the conclusion of docker run when the container dies, but I put logic there to remove the file in case.

### CODE REVIEW

File: app/spreadsheet.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    s3_client = boto3.client('s3')
+    bucket = "nagbot-spreadhseets"
+    try:
+        # response = s3_client.upload_file(file_name, bucket, object_name)
+        s3_client.upload_file(filename, bucket, filename)
+    except ClientError as e:
+        print(e)
+
+    # Cleanup and delete the workbook
+    os.remove(filename)
```
Comment (alizbay94) [inline app/spreadsheet.py]: wasn't sure if I needed this since it will be running in a docker container and file should be gone when container dies, but left there for good measure.

File: app/nagbot.py [hunk 2]
```diff
@@ -28,7 +29,7 @@
 3. PIP dependencies specified in requirements.txt.
 4. Environment variables
    * "SLACK_BOT_TOKEN" containing a token allowing messages to be posted to Slack.
-   * "GDOCS_SERVICE_ACCOUNT_FILENAME" containing the name of the google sheet
+   * "GDOCS_SERVICE_ACCOUNT_FILENAME" containing the name of the google sheet
```
Comment (CodyRay) [inline app/nagbot.py]: Is all the Gdocs stuff ripped out? Maybe we can remove these references

File: app/spreadsheet.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    # create header in worksheet
+    bold_format = workbook.add_format({'bold': True, "align": "left", "font_size": 15})
+    standard_format = workbook.add_format({"font_size": 12, 'align': 'left'})
+    worksheet.write_row(0, 0, resource_header, bold_format)
+
+    # widen the columns for readability
+    worksheet.set_column(0, len(resource_header)+1, 20)
+
+    # Write resources data to worksheet
+    col_num = 1
```
Comment (CodyRay) [inline app/spreadsheet.py]: Consider using `enumerate()` https://www.geeksforgeeks.org/enumerate-in-python/

File: app/spreadsheet.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+def upload_spreadsheet_to_s3(filename, workbook):
+    # save the workbook to the cwd by closing
+    workbook.close()
+
+    # Upload the file to S3
+    s3_client = boto3.client('s3')
+    bucket = "nagbot-spreadhseets"
+    try:
+        # response = s3_client.upload_file(file_name, bucket, object_name)
```
Comment (CodyRay) [inline app/spreadsheet.py]: Commented code could be deleted :recycle:

File: app/nagbot.py [hunk 5]
```diff
@@ -38,9 +39,15 @@ def notify_internal(channel, dryrun):
         summary_msg = f"Hi, I'm Nagbot v{__version__} :wink: "
         summary_msg += "My job is to make sure we don't forget about unwanted AWS resources and waste money!\n"
 
+        filename = f"{TODAY_YYYY_MM_DD}-NagBot-Report.xlsx"
```
Comment (CodyRay) [inline app/nagbot.py:39]: Consider using [TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory), then python will take care of cleaning up and it is in a location that is generally writable by all users

File: app/nagbot.py [hunk 6]
```diff
@@ -81,6 +88,7 @@ def notify_internal(channel, dryrun):
                                          dryrun=dryrun)
                 else:
                     summary_msg += f'No {ec2_type}s are due to be stopped at this time.\n'
+        spreadsheet.upload_spreadsheet_to_s3(filename, workbook)
         sqslack.send_message(channel, summary_msg)
```
Comment (CodyRay) [inline app/nagbot.py:94]: Can we link the url in the message?

File: app/spreadsheet.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    # Write resources data to worksheet
+    col_num = 1
+    for resource in resources:
+        resource_data = resource.to_list()
+        # convert any lists in resource_data from list to formatted string
+        resource_data = [", ".join(data) if type(data) is list else data for data in resource_data]
+        worksheet.write_row(col_num, 0, resource_data, standard_format)
+        col_num += 1
+    return workbook
```
Comment (CodyRay) [inline app/spreadsheet.py:41]: Doing something like https://stackoverflow.com/questions/29463274/simulate-autofit-column-in-xslxwriter would be nice to have; make it a little bit nicer to use

If we could turn on the [make it a table](https://www.tutorialspoint.com/python_xlsxwriter/python_xlsxwriter_tables.htm) that would be handy too

**Review by CodyRay — COMMENTED** [review]
Looking great, some comments an improvements

File: app/nagbot.py (hunk 2 above)
Comment (alizbay94) [reply app/nagbot.py]: I think its all gone, i'll double check and remove the old references.

File: app/nagbot.py (hunk 6 above)
Comment (alizbay94) [reply app/nagbot.py:94]: Good idea, I could set that up.

File: app/spreadsheet.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    return xlsxwriter.Workbook(filename)
+
+
+def add_resource_worksheet_to_workbook(workbook, resources, resource_name):
+    worksheet = workbook.add_worksheet(resource_name)
+    resource_header = resources[0].to_header()
+
+    # create header in worksheet
+    bold_format = workbook.add_format({'bold': True, "align": "left", "font_size": 15})
+    standard_format = workbook.add_format({"font_size": 12, 'align': 'left'})
```
Comment (CodyRay) [inline app/spreadsheet.py]: I'd vote to remove the font size configuration - seems very big

**Review by CodyRay — COMMENTED** [review]
Some further notes after playing with the example workbook:

I noticed some weirdness for Snapshots: they have some columns - OS, IOPS, and Throughput that don't make sense for volumes

Some other ideas for useability improvements:
 * Could the ID columns be clickable links to open the resource in the console?
 * Could we add an "Overview" worksheet that has a summary of spending for each resource type?
 * Could we sort the rows by descending costs? Have the highest cost items front and center

If any of those nice to haves end up being difficult to add, that is ok to skip them

File: app/spreadsheet.py [hunk 9]
```diff
@@ -32,17 +33,19 @@ def add_resource_worksheet_to_workbook(workbook, resources, resource_name):
 
 
 def upload_spreadsheet_to_s3(filename, workbook):
-    # save the workbook to the cwd by closing
+    cwd = os.getcwd()
+    # save the workbook to a temporary directory
+    temp_directory = tempfile.TemporaryDirectory(ignore_cleanup_errors=True)
```
Comment (CodyRay) [inline app/spreadsheet.py]: Does this imply the file isn't actually created until the `close` command is called? If so, I think you can revert the changes to the Dockerfile

Comment (CodyRay) [reply app/spreadsheet.py]: Consider [with](https://realpython.com/python-with-statement/) or a try-finally to ensure cleanup

**Review by CodyRay — APPROVED** [review]

**Comment by alizbay94** [comment]
I've implemented logic to include hyperlinks for each resource id in the excel file that links to the resource on AWS. I've also made the data from each resource worksheet into a table, and wrote custom logic to format the column width of each column based on the longest string in the column including the header. 
I'm thinking I won't sort the table entries in descending order of price since it would require writing custom logic, and we could just use the table to sort the entries in any order we choose when we open the excel file in the Excel UI. There's a drop down next to each column header where we can sort the entries however we choose.

The only feedback i've got left to address in this PR is adding an 'Overview' sheet - @haroldhues, would the idea be that this sheet would essentially contain 4 entries, one entry for the total cost for each of the resources?

File: app/spreadsheet.py (hunk 9 above)
Comment (alizbay94) [reply app/spreadsheet.py]: Implementing a try-finally now

**Review by CodyRay — APPROVED** [review]
Nice! Great improvements

File: app/snapshot.py [hunk 10]
```diff
@@ -31,10 +32,7 @@ def to_header() -> [str]:
                 'Monthly Price',
```
Comment (CodyRay) [inline app/snapshot.py:34]: I found that the spreadsheet doesn't sort the price columns for instances correctly because they contain a literal dollar sign.

image

I think the contents of the cell needs to be without the dollar sign and then format as "Currency". Then it will appear with the dollar sign, but Excel will do calculations without the dollar sign.

This only seems to apply to instances, the other types don't have the dollar sign.

File: app/ami.py [hunk 11]
```diff
@@ -32,9 +32,8 @@ def to_header() -> [str]:
                 'AMI Type',
                 'OS',
                 'IOPS',
-                'VolumeType'
-                'Throughput',
-                'Snapshot IDs']
+                'VolumeType',
```
Comment (CodyRay) [inline app/ami.py:37]: I think for AMI we are missing the "AMI name" . Every AMI seems to have it, but I think we are pulling the Name tag. For example, https://us-west-1.console.aws.amazon.com/ec2/v2/home?region=us-west-1#Images:visibility=owned-by-me;search=ami-046531e848f15f786 doesn't show a name in the spreadsheet

Also, do we know if we have creation date available for all resources?

**Review by CodyRay — APPROVED** [review]

**Comment by CodyRay** [comment]
> The only feedback i've got left to address in this PR is adding an 'Overview' sheet - @haroldhues, would the idea be that this sheet would essentially contain 4 entries, one entry for the total cost for each of the resources?

Yep, I think we could potential show other things there in the future as well like a regional breakdown or counts of resources, but total cost for each resource is a good place to start imo

File: app/snapshot.py (hunk 10 above)
Comment (alizbay94) [reply app/snapshot.py:34]: I've noticed the same thing, it must be due to the way instance prices are calculated. I'll get them all uniform!

**Comment by alizbay94** [comment]
@haroldhue , each of the resources has a creation timestamp, but it is called something different for each resource. I implemented the logic to retrieve the creation timestamp for each resource depending on what it is called, and added an attribute to each resource object with that info, and included a column with that info for each worksheet.
I also implemented the logic to retrieve AMI names, and the AMI names are showing up as expected now 👍🏼
Finally I created an additional 'Summary' worksheet in the workbook with the total cost of each resource.

**Review by CodyRay — APPROVED** [review]

**Comment by CodyRay** [comment]
Latest version is working pretty well - well enough for me to get myself into trouble :cry: 

Is it possible to make the summary the first worksheet?

**Comment by alizbay94** [comment]
Yup! I can get that done.

**Comment by alizbay94** [comment]
I was able to find a method of the `worksheet` class called `activate()` which enables us to set the summary sheet as the first visible sheet when the workbook is open, i've included it in the `add_summary_worksheet_to_workbook` method.

File: app/spreadsheet.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        try:
+            s3_client.upload_file(filename, bucket, filename)
+        except ClientError as e:
+            print(e)
+
+    finally:
+        os.chdir(cwd)
+        temp_directory.cleanup()
+
+    return f"https://s3.console.aws.amazon.com/s3/buckets/{bucket}"
```
Comment (CodyRay) [inline app/spreadsheet.py]: I think if you add `?prefix={filename}` here then you'll get a link directly to the uploaded file

**Review by CodyRay — APPROVED** [review]
One last suggestion, but working great - ship it!

File: app/spreadsheet.py (hunk 12 above)
Comment (alizbay94) [reply app/spreadsheet.py]: I'll add that and test before I merge 👍🏼
